src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs

```diff
@@ -64,21 +75,68 @@ public Context CreateContext(IFeatureCollection contextFeatures)
             }
-
             if (diagnoticsEnabled)
             {
+#if !NET451
+                context.Activity = StartActivity(httpContext);
+#endif
                 // Non-inline
                 RecordBeginRequestDiagnostics(httpContext, startTimestamp);
             }
 
-            // Create and return the request Context
-            return new Context
+            // Fill and return the request Context
+            context.HttpContext = httpContext;
+            context.Scope = scope;
+            context.StartTimestamp = startTimestamp;
+            return context;
+        }
+
+#if !NET451
+        private Activity StartActivity(HttpContext httpContext)
+        {
+            var activity = new Activity(ActivityName);
+            StringValues requestId;
+            if (httpContext.Request.Headers.TryGetValue(_activityTrackingOptions.RequestIdHeaderName, out requestId))
+            {
+                try
+                {
+                    activity.SetParentId(requestId);
+                }
+                catch (ArgumentException ex)
+                {
+                    _logger.LogWarning(LoggerEventIds.BadRequestIdHeader, ex, "Request ID received in '{RequestIdHeaderName}' header is invalid '{RequestId}'", _activityTrackingOptions.RequestIdHeaderName, requestId);
+                }
+
+                // We expect baggage to be empty by default
+                // Only very advanced users will be using it in near future, we encouradge them to keep baggage small (few items)
+                string[] baggage = httpContext.Request.Headers.GetCommaSeparatedValues(_activityTrackingOptions.BaggageHeaderName);
+                if (baggage != StringValues.Empty)
+                {
+                    foreach (var item in baggage)
+                    {
+                        NameValueHeaderValue baggageItem;
+                        if (NameValueHeaderValue.TryParse(item, out baggageItem))
+                        {
+                            try
+                            {
+                                activity.AddBaggage(baggageItem.Name, baggageItem.Value);
+                            }
+                            catch (ArgumentException ex)
+                            {
+                                _logger.LogWarning(LoggerEventIds.BadRequestIdHeader, ex, "Baggage item received in '{BaggageHeaderName}' header is invalid '{ItemName}' with value '{ItemValue}'", _activityTrackingOptions.BaggageHeaderName, baggageItem.Name, baggageItem.Value);
+                            }
+                        }
+                    }
+                }
+            }
+            if (_diagnosticSource.IsEnabled(DiagnosticsBeginRequestKey, activity, httpContext))
             {
-                HttpContext = httpContext,
-                Scope = scope,
-                StartTimestamp = startTimestamp,
-            };
+                return null;
+            }
+            activity.Start();
```

Comment:
Why doesn't this use the `StartActivity` extension method to also raise the corresponding event with access to properties like the HttpContext? Aren't the old `BeginRequest`/`EndRequest` events legacy now - similar to what you've done with HttpRequest?

Same for the `StopActivity` extension method.